app/javascript/components/bookmarks/BookmarkManager.jsx

```diff
@@ -0,0 +1,286 @@
+    setCurrentBookmark(prevBookmark => {
+      const newFormState = _cloneDeep(prevBookmark)
+      Object.assign(newFormState, update)
+      return newFormState
+    })
+  }
+
+  /** set state/text on Save button */
+  function enableSaveButton(enabled) {
+    if (enabled) {
+      setSaveText('Save')
+      setSaveDisabled(null)
+    } else {
+      setSaveText('Saving...')
+      setSaveDisabled(true)
+    }
+  }
+
+  /** show contextual errors */
+  function handleErrorContent(error) {
+    const errorMessage = formatErrorMessages(error)
+    setError(errorMessage)
+    setShowError(true)
+  }
+
+  /** format errors object into comma-delimited message */
+  function formatErrorMessages(error) {
+    if (error.errors) {
+      return Object.keys(error.errors).map(key => {
+        return `${key} ${errors[key][0]}`
+      }).join(', ')
+    } else {
+      return error.message
+    }
+  }
+
+  /** create/update a new bookmark */
+  async function handleSaveBookmark(e) {
+    e.preventDefault()
+    enableSaveButton(false)
+    const saveProps = [currentBookmark]
+    const saveFunc = bookmarkSaved ? updateBookmark : createBookmark
+    if (bookmarkSaved) {
+      saveProps.unshift(getBookmarkId(currentBookmark))
+    }
+    try {
+      const bookmark = await saveFunc(...saveProps)
+      enableSaveButton(true)
+      setShowError(false)
+      resetForm()
+      updateBookmarkList(bookmark)
+    } catch (error) {
+      enableSaveButton(true)
+      handleErrorContent(error)
+    }
+  }
+
+  /** delete a bookmark */
+  async function handleDeleteBookmark(e) {
+    e.preventDefault()
+    setDeleteDisabled(true)
+    const toDelete = getBookmarkId(currentBookmark)
+    if (!toDelete) {
+      setDeleteDisabled(false)
+      return false
+    } else {
+      try {
+        await deleteBookmark(toDelete)
+        setDeleteDisabled(false)
+        setShowError(false)
+        removeBookmarkFromList(toDelete)
+      } catch (error) {
+        setDeleteDisabled(false)
+        handleErrorContent(error)
+      }
+    }
+  }
+
+  const [serverBookmarks, setServerBookmarks] = useState([])
+  const [serverBookmarksLoaded, setServerBookmarksLoaded] = useState(false)
+  const [showBookmarksModal, setShowBookmarksModal] = useState(false)
+
+  const toggleBookmarkModal = () => {
+    setShowBookmarksModal(!showBookmarksModal)
+  }
+
+  /** load all user bookmarks from server */
+  async function loadServerBookmarks() {
+    toggleBookmarkModal()
+    if (!serverBookmarksLoaded) {
+      try {
+        const serverUserBookmarks = await fetchBookmarks()
+        setServerBookmarks(serverUserBookmarks)
+        setServerBookmarksLoaded(true)
+      } catch (error) {
+        setShowBookmarksModal(false)
+        setServerBookmarks([])
+        setServerBookmarksLoaded(false)
+      }
+    }
+  }
+
+  const loginPopover = <Popover data-analytics-name='login-bookmark-popover' id='login-bookmark-popover'>
+    You must sign in to bookmark this view
```

Comment:
Non-blocking: to lower interaction cost and thus boost engagement, can we make this easier for the user like so?

```suggestion
    Click to sign in, then bookmark this view
```

In other words, clicking the star would bring the user to the sign-in view, then they'd sign in, then they'd be brought back here, then they could click to bookmark.  It avoids needing to find the "Sign in" button (which new users might not notice, even though it's prominent) and clicking it.

That elides some detail of course, but such copy keeps it light and I think users would get it.  We could perhaps pave this road more later -- by e.g. enabling bookmarks via a fancy callback URL so the bookmark would be set automatically, with a transient "Saved bookmark" tooltip on page load -- but I suspect even this initial lower-interaction-cost UX would help more users would engage with bookmarks.

Reply:
Makes sense - good refinement.